### PR TITLE
fix(query): the hashmap will cause oom

### LINKED ISSUE #8615: bug: col1 like col2 will be oom kill
**Summary**

table 15000000 rows

```sql
databend> desc orders;
+-----------------+---------+------+---------+-------+
| Field           | Type    | Null | Default | Extra |
+-----------------+---------+------+---------+-------+
| o_orderkey      | BIGINT  | NO   | 0       |       |
| o_custkey       | BIGINT  | NO   | 0       |       |
| o_orderstatus   | VARCHAR | NO   |         |       |
| o_totalprice    | DOUBLE  | NO   | 0       |       |
| o_orderdate     | DATE    | NO   | 0       |       |
| o_orderpriority | VARCHAR | NO   |         |       |
| o_clerk         | VARCHAR | NO   |         |       |
| o_shippriority  | INT     | NO   | 0       |       |
| o_comment       | VARCHAR | NO   |         |       |
+-----------------+---------+------+---------+-------+

```

execute query on databend-query release:

```sql
select count() from orders where o_clerk like o_comment;
```

in /var/log/syslog

```
26811 Nov  3 10:26:26  kernel: [50984.173571] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=c7d8e9fa439db04e33d5cd92d1330ed38d965c345e2d396d7f3ba920c5b65bfa,mems_allowed=0,global_oom,task_memcg=/user.slice/user-1000.slice/user@1000.service,task=databen      d-query,pid=170805,uid=1000
26812 Nov  3 10:26:26  kernel: [50984.173640] Out of memory: Killed process 170805 (databend-query) total-vm:23283508kB, anon-rss:11226736kB, file-rss:0kB, shmem-rss:0kB, UID:1000 pgtables:33288kB oom_score_adj:0
```

execute query on ck :

```sql
ck :)  select count() from orders where o_clerk like o_comment;

SELECT count()
FROM orders
WHERE o_clerk LIKE o_comment

Query id: ee2c5d3e-b4c4-44ed-9332-9eb04c274e0c

┌─count()─┐
│       0 │
└─────────┘

1 row in set. Elapsed: 14.368 sec. Processed 15.00 million rows, 1.22 GB (1.04 million rows/s., 84.80 MB/s.)

```

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Now `l_col like r_col` will generate a hashmap based on r_col, if r_col is huge, it will be oom kill.

sys environment:

> OS: Ubuntu 22.04.1 LTS (Jammy Jellyfish)
> CPU: Intel(R) Xeon(R) Platinum 8269CY CPU @ 2.50GHz
> Memory: 30G
> Disk: 16G / 197G

### Table Struct

```sql
-- databend
 CREATE TABLE `orders` (
  `o_orderkey` BIGINT,
  `o_custkey` BIGINT,
  `o_orderstatus` VARCHAR,
  `o_totalprice` DOUBLE,
  `o_orderdate` DATE,
  `o_orderpriority` VARCHAR,
  `o_clerk` VARCHAR,
  `o_shippriority` INT,
  `o_comment` VARCHAR
)
-- ck
SELECT version()

Query id: 73c60cca-5362-4765-b4fa-9e0a7a9cdd8b

┌─version()──┐
│ 22.10.2.11 │
└────────────┘

CREATE TABLE tpch.orders
(
    `o_orderkey` Int64,
    `o_custkey` Int32,
    `o_orderstatus` String,
    `o_totalprice` Decimal(15, 2),
    `o_orderdate` Date,
    `o_orderpriority` String,
    `o_clerk` String,
    `o_shippriority` Int32,
    `o_comment` String
)
ENGINE = MergeTree
ORDER BY o_orderkey
SETTINGS index_granularity = 8192

```

### Prepare Data

data from: https://github.com/leiysky/tpch-databend 

```shell
./tpch.sh 10.0
bash insert_data.sh
```

```sql
-- databend
mysql> select count() from orders ;
+----------+
| count()  |
+----------+
| 15000000 |
+----------+
1 row in set (0.01 sec)
Read 1 rows, 1.00 B in 0.001 sec., 796.38 rows/sec., 796.38 B/sec.


-- ck
SELECT count()
FROM orders

Query id: 173ca818-69f0-4935-8526-429cb321eac5

┌──count()─┐
│ 15000000 │
└──────────┘

1 row in set. Elapsed: 0.001 sec. 

```

### Result

test sql: `select count() from orders where o_clerk not  like o_comment;`

| before optimize | drop hash map | after optimize | ck |
|--|--|--|--|
| oom |114s|0.832 s|7.925s|

Closes #8615

### CODE REVIEW

**Review by sundy-li — APPROVED** [review]

**Comment by TCeason** [comment]
ali cloud 16C32G
all settings is default.

ck    databend
7.9s  114s

```sql
-- databend
mysql> select count() from orders where o_clerk not  like o_comment;
+----------+
| count()  |
+----------+
| 15000000 |
+----------+
1 row in set (1 min 53.77 sec)
Read 15000000 rows, 1.11 GiB in 113.761 sec., 131.86 thousand rows/sec., 10.00 MiB/sec.


-- ck


SELECT version()

Query id: 58b06879-7ca4-4854-96d1-d4dd02cc2540

┌─version()──┐
│ 22.10.2.11 │
└────────────┘


SELECT count()
FROM orders
WHERE o_clerk NOT LIKE o_comment

Query id: 0d1ff665-04e1-435d-bbb7-e21da85dff00

┌──count()─┐
│ 15000000 │
└──────────┘

1 row in set. Elapsed: 7.925 sec. Processed 15.00 million rows, 1.22 GB (1.89 million rows/s., 153.73 MB/s.)


```

**Comment by Xuanwo** [comment]
How does clickhouse handle this?

**Comment by Xuanwo** [comment]
BTW, like pattern is much simpler than regex. Maybe we don't need to convert like pattern into regex?

**Comment by Xuanwo** [comment]
> BTW, like pattern is much simpler than regex. Maybe we don't need to convert like pattern into regex?

I port a quick demo from [tikv](https://github.com/tikv/tikv/blob/master/components/tidb_query_expr/src/impl_like.rs)

The benches show that plain like is much faster than regex build (if we have to):

```shell
regex/build_regex       time:   [11.830 µs 11.857 µs 11.902 µs]
                        change: [-5.3625% -4.7534% -4.1822%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  8 (8.00%) low mild
  3 (3.00%) high severe
regex/build_like        time:   [3.1063 ns 3.1160 ns 3.1277 ns]
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe
```

<details>
<summary>The source code</summary>

```rust
fn bench_regex(c: &mut Criterion) {
    let mut group = c.benchmark_group("regex");

    group.bench_function("build_regex", |b| {
        b.iter(|| {
            let _ = regex::Regex::new("^Hello\\..World.*%$");
        })
    });
    group.bench_function("build_like", |b| {
        b.iter(|| {
            let _ = like(b"Hello, World", b"^Hello\\..World.*%$", &0);
        })
    });
}

fn decode_one(data: &[u8]) -> Option<(u8, usize)> {
    if data.is_empty() {
        None
    } else {
        Some((data[0], 1))
    }
}

fn sort_compare(a: &[u8], b: &[u8]) -> std::cmp::Ordering {
    a.cmp(b)
}

fn like(target: &[u8], pattern: &[u8], escape: &i64) -> Option<i64> {
    let escape = *escape as u32;
    // current search positions in pattern and target.
    let (mut px, mut tx) = (0, 0);
    // positions for backtrace.
    let (mut next_px, mut next_tx) = (0, 0);
    while px < pattern.len() || tx < target.len() {
        if let Some((c, mut poff)) = decode_one(&pattern[px..]) {
            let code: u32 = c.into();
            if code == '_' as u32 {
                if let Some((_, toff)) = decode_one(&target[tx..]) {
                    px += poff;
                    tx += toff;
                    continue;
                }
            } else if code == '%' as u32 {
                // update the backtrace point.
                next_px = px;
                px += poff;
                next_tx = tx;
                next_tx += if let Some((_, toff)) = decode_one(&target[tx..]) {
                    toff
                } else {
                    1
                };
                continue;
            } else {
                if code == escape && px + poff < pattern.len() {
                    px += poff;
                    poff = if let Some((_, off)) = decode_one(&pattern[px..]) {
                        off
                    } else {
                        break;
                    }
                }
                if let Some((_, toff)) = decode_one(&target[tx..]) {
                    if let std::cmp::Ordering::Equal =
                        sort_compare(&target[tx..tx + toff], &pattern[px..px + poff])
                    {
                        tx += toff;
                        px += poff;
                        continue;
                    }
                }
            }
        }
        // mismatch and backtrace to last %.
        if 0 < next_tx && next_tx <= target.len() {
            px = next_px;
            tx = next_tx;
            continue;
        }
        return Some(false as i64);
    }

    Some(true as i64)
}
```

</detials>

**Comment by BohuTANG** [comment]
This case is not a constant regex rule,  it's column-1 like column-2, is more difficult here.

**Comment by TCeason** [comment]
> > BTW, like pattern is much simpler than regex. Maybe we don't need to convert like pattern into regex?
> 
> I port a quick demo from [tikv](https://github.com/tikv/tikv/blob/master/components/tidb_query_expr/src/impl_like.rs)
> 
> Conclusion: Regex matching is `O(n)`, but the regex build is not. If we need to build regex every time, it's better to implement like directly. (And in some simple case, like is just faster than regex)
> 
> The benches show that plain like is much faster than regex build (if we have to):
> 
> ```shell
> regex/build_regex_everytime
>                         time:   [13.348 µs 13.371 µs 13.392 µs]
> regex/reuse_regex       time:   [16.677 ns 16.699 ns 16.722 ns]
> regex/like              time:   [3.0577 ns 3.0593 ns 3.0611 ns]
> ```
> 
> Note: the like function is ported here just for demo. It could be wrong.
> 
> The source code
> ```rust
> fn bench_regex(c: &mut Criterion) {
>     let mut group = c.benchmark_group("regex");
> 
>     group.bench_function("build_regex_everytime", |b| {
>         b.iter(|| {
>             let re = regex::Regex::new("^Hello\\..World.*%$").unwrap();
>             re.is_match("Hello, World!");
>         })
>     });
> 
>     let built_re = regex::Regex::new("^Hello\\..World.*%$").unwrap();
>     group.bench_function("reuse_regex", |b| {
>         b.iter(|| {
>             built_re.is_match("Hello, World!");
>         })
>     });
>     group.bench_function("like", |b| {
>         b.iter(|| {
>             let _ = like(b"Hello, World", b"^Hello\\..World.*%$", &0);
>         })
>     });
> }
> 
> fn decode_one(data: &[u8]) -> Option<(u8, usize)> {
>     if data.is_empty() {
>         None
>     } else {
>         Some((data[0], 1))
>     }
> }
> 
> fn sort_compare(a: &[u8], b: &[u8]) -> std::cmp::Ordering {
>     a.cmp(b)
> }
> 
> fn like(target: &[u8], pattern: &[u8], escape: &i64) -> Option<i64> {
>     let escape = *escape as u32;
>     // current search positions in pattern and target.
>     let (mut px, mut tx) = (0, 0);
>     // positions for backtrace.
>     let (mut next_px, mut next_tx) = (0, 0);
>     while px < pattern.len() || tx < target.len() {
>         if let Some((c, mut poff)) = decode_one(&pattern[px..]) {
>             let code: u32 = c.into();
>             if code == '_' as u32 {
>                 if let Some((_, toff)) = decode_one(&target[tx..]) {
>                     px += poff;
>                     tx += toff;
>                     continue;
>                 }
>             } else if code == '%' as u32 {
>                 // update the backtrace point.
>                 next_px = px;
>                 px += poff;
>                 next_tx = tx;
>                 next_tx += if let Some((_, toff)) = decode_one(&target[tx..]) {
>                     toff
>                 } else {
>                     1
>                 };
>                 continue;
>             } else {
>                 if code == escape && px + poff < pattern.len() {
>                     px += poff;
>                     poff = if let Some((_, off)) = decode_one(&pattern[px..]) {
>                         off
>                     } else {
>                         break;
>                     }
>                 }
>                 if let Some((_, toff)) = decode_one(&target[tx..]) {
>                     if let std::cmp::Ordering::Equal =
>                         sort_compare(&target[tx..tx + toff], &pattern[px..px + poff])
>                     {
>                         tx += toff;
>                         px += poff;
>                         continue;
>                     }
>                 }
>             }
>         }
>         // mismatch and backtrace to last %.
>         if 0 < next_tx && next_tx <= target.len() {
>             px = next_px;
>             tx = next_tx;
>             continue;
>         }
>         return Some(false as i64);
>     }
> 
>     Some(true as i64)
> }
> ```

Yes our vector_const is a same impl like this.

https://github.com/datafuselabs/databend/blob/3cd696a195431903d921cf3bb17461f6e99faf1f/src/query/functions/src/scalars/comparisons/comparison_like.rs#L67

**Comment by sundy-li** [comment]
```
fn like(target: &[u8], pattern: &[u8], escape: &i64) -> Option<i64> {
```

The codes are better than the regex version, I vote to have this change.

**Comment by TCeason** [comment]
> ```
> fn like(target: &[u8], pattern: &[u8], escape: &i64) -> Option<i64> {
> ```
> 
> 
>      
>   
> 
> The codes are better than the regex version, I vote to have this change.  cc @TCeason   It's ok to address this improvement in this pr.

Ok

File: src/query/functions/src/scalars/comparisons/comparison_like.rs [hunk 1]
```diff
@@ -229,3 +209,77 @@ pub fn like_pattern_to_regex(pattern: &str) -> String {
...
+
+fn decode_one(data: &[u8]) -> Option<(u8, usize)> {
+    if data.is_empty() {
+        None
+    } else {
+        Some((data[0], 1))
+    }
+}
+
+// Borrow from tikv/tikv/blob/master/components/tidb_query_expr/src/impl_like.rs
```
Comment (Xuanwo) [inline src/query/functions/src/scalars/comparisons/comparison_like.rs:230]:
```suggestion
/// Borrow from [tikv](https://github.com/tikv/tikv/blob/fe997db4db8a5a096f8a45c0db3eb3c2e5879262/components/tidb_query_expr/src/impl_like.rs)
```

**Comment by TCeason** [comment]
If the like pattern just has one substring, we keep the original logic.

```
regex/build_regex_everytime                                                                             
                        time:   [29.039 µs 29.068 µs 29.098 µs]
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe
regex/reuse_regex       time:   [46.815 ns 46.837 ns 46.861 ns]                               
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe
regex/like              time:   [41.852 ns 41.869 ns 41.886 ns]                        
Found 15 outliers among 100 measurements (15.00%)
  14 (14.00%) high mild
  1 (1.00%) high severe
regex/match_one_pattern time:   [0.0000 ps 0.0000 ps 0.0000 ps]                                             
Found 12 outliers among 100 measurements (12.00%)
  4 (4.00%) high mild
  8 (8.00%) high severe
```

<details>
<summary>The source code</summary>

```shell
~/hello# tree 
.
├── Cargo.lock
├── Cargo.toml
├── benches
│   └── benchmark.rs
└── src
    └── main.rs

2 directories, 4 files

```

```rust
~/hello# cat Cargo.toml 
[package]
name = "hello"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
memchr = "2.5.0"
regex = "1.6.0"
criterion = "0.3"

[[bench]]
name = "benchmark"
harness = false

~/hello# cat benches/benchmark.rs 
use criterion::{criterion_group, criterion_main, Criterion};

fn decode_one(data: &[u8]) -> Option<(u8, usize)> {
    if data.is_empty() {
        None
    } else {
        Some((data[0], 1))
    }
}

fn like(target: &[u8], pattern: &[u8]) -> bool {
    // current search positions in pattern and target.
    let (mut px, mut tx) = (0, 0);
    // positions for backtrace.
    let (mut next_px, mut next_tx) = (0, 0);
    while px < pattern.len() || tx < target.len() {
        if let Some((c, mut poff)) = decode_one(&pattern[px..]) {
            let code: u32 = c.into();
            if code == '_' as u32 {
                if let Some((_, toff)) = decode_one(&target[tx..]) {
                    px += poff;
                    tx += toff;
                    continue;
                }
            } else if code == '%' as u32 {
                // update the backtrace point.
                next_px = px;
                px += poff;
                next_tx = tx;
                next_tx += if let Some((_, toff)) = decode_one(&target[tx..]) {
                    toff
                } else {
                    1
                };
                continue;
            } else {
                if code == 0 && px + poff < pattern.len() {
                    px += poff;
                    poff = if let Some((_, off)) = decode_one(&pattern[px..]) {
                        off
                    } else {
                        break;
                    }
                }
                if let Some((_, toff)) = decode_one(&target[tx..]) {
                    if let std::cmp::Ordering::Equal = target[tx..tx + toff].cmp(&pattern[px..px + poff])
                    {
                        tx += toff;
                        px += poff;
                        continue;
                    }
                }
            }
        }
        // mismatch and backtrace to last %.
        if 0 < next_tx && next_tx <= target.len() {
            px = next_px;
            tx = next_tx;
            continue;
        }
        return false;
    }
	true
}

fn search_sub_str(str: &[u8], substr: &[u8]) -> Option<usize> {
    if substr.len() <= str.len() {
        str.windows(substr.len()).position(|w| w == substr)
    } else {
        None
    }
}

fn bench_regex(c: &mut Criterion) {
    let mut group = c.benchmark_group("regex");

    group.bench_function("build_regex_everytime", |b| {
        b.iter(|| {
            let re = regex::Regex::new("^.*Hello.*$").unwrap();
            re.is_match("Hello, World!");
        })
    });

    let built_re = regex::Regex::new("^.*Hello.*$").unwrap();
    group.bench_function("reuse_regex", |b| {
        b.iter(|| {
            built_re.is_match("Hello, World!");
        })
    });
    group.bench_function("like", |b| {
        b.iter(|| {
            if like(b"Hello, World", b"%Hello%") {
            } else {
            }
        })
    });
    group.bench_function("match_one_pattern", |b| {
        b.iter(|| {
            if search_sub_str(b"Hello, World", b"Hello").is_some() {} else {}
        })
    });
}

criterion_group!(benches, bench_regex);
criterion_main!(benches);


```
</details>

**Comment by TCeason** [comment]
Ready for review. cc @sundy-li  @BohuTANG  @Xuanwo

File: src/query/functions-v2/tests/it/scalars/testdata/comparison.txt [hunk 2]
```diff
@@ -885,15 +903,15 @@ evaluation:
 | Row 0  | "abc"           | "a%"            | true    |
 | Row 1  | "abd"           | "_b_"           | true    |
 | Row 2  | "abe"           | "abe"           | true    |
-| Row 3  | "abf"           | "a"             | false   |
+| Row 3  | "abf"           | "a"             | true    |
```
Comment (TCeason) [inline src/query/functions-v2/tests/it/scalars/testdata/comparison.txt:906]: maybe a bug

**Comment by Xuanwo** [comment]
> ```sql
> mysql> select count() from orders where o_clerk not  like o_comment;
> ```

Can we have a benchmark like this again?

**Comment by TCeason** [comment]
> > ```sql
> > mysql> select count() from orders where o_clerk not  like o_comment;
> > ```
> 
> Can we have a benchmark like this again?

Speed up.
start with command: ./target/release/databend-query 

```sql
Database changed
--first exec
mysql> select count() from orders where o_clerk not  like o_comment;
+----------+
| count()  |
+----------+
| 15000000 |
+----------+
1 row in set (1.66 sec)
Read 15000000 rows, 1.11 GiB in 1.656 sec., 9.06 million rows/sec., 686.79 MiB/sec.

--second exec
mysql> select count() from orders where o_clerk not  like o_comment;
+----------+
| count()  |
+----------+
| 15000000 |
+----------+
1 row in set (0.84 sec)
Read 15000000 rows, 1.11 GiB in 0.832 sec., 18.03 million rows/sec., 1.33 GiB/sec.


```

**Comment by TCeason** [comment]
> @TCeason
> 
> I prefer to edit the PR summary and give the performance result comparison. For example: #8452

Done

**Pull request merged**